winch/codegen/src/isa/aarch64/masm.rs

```diff
@@ -1,57 +1,178 @@
     }
 
-    fn local_address(&mut self, _local: &LocalSlot) -> Address {
-        todo!()
+    fn local_address(&mut self, local: &LocalSlot) -> Address {
+        let (reg, offset) = local
+            .addressed_from_sp()
+            .then(|| {
+                let offset = self.sp_offset.checked_sub(local.offset).expect(&format!(
+                    "Invalid local offset = {}; sp offset = {}",
+                    local.offset, self.sp_offset
+                ));
+                (regs::shadow_sp(), offset)
+            })
+            .unwrap_or((regs::fp(), local.offset));
+
+        Address::offset(reg, offset as i64)
     }
 
-    fn store(&mut self, _src: RegImm, _dst: Address, _size: OperandSize) {
-        todo!()
+    fn store(&mut self, src: RegImm, dst: Address, size: OperandSize) {
+        let src = match src {
+            RegImm::Imm(imm) => {
+                let scratch = regs::scratch();
+                self.asm.load_constant(imm as u64, scratch);
+                scratch
+            }
+            RegImm::Reg(reg) => reg,
+        };
+
+        self.asm.str(src, dst, size);
     }
 
-    fn load(&mut self, _src: Address, _dst: Reg, _size: OperandSize) {}
+    fn load(&mut self, src: Address, dst: Reg, size: OperandSize) {
+        self.asm.ldr(src, dst, size);
+    }
 
     fn sp_offset(&mut self) -> u32 {
-        0u32
+        self.sp_offset
     }
 
     fn finalize(self) -> MachBufferFinalized<Final> {
-        todo!()
+        self.asm.finalize()
+    }
+
+    fn mov(&mut self, src: RegImm, dst: RegImm, size: OperandSize) {
+        self.asm.mov(src.into(), dst.into(), size);
+    }
+
+    fn add(&mut self, dst: RegImm, lhs: RegImm, rhs: RegImm, size: OperandSize) {
+        self.asm.add(rhs.into(), lhs.into(), dst.into(), size);
     }
 
-    fn mov(&mut self, _src: RegImm, _dst: RegImm, _size: OperandSize) {
-        todo!()
+    fn zero(&mut self, reg: Reg) {
+        self.asm.load_constant(0, reg);
     }
 
-    fn add(&mut self, _dst: RegImm, __lhs: RegImm, __rhs: RegImm, _size: OperandSize) {
-        todo!()
+    fn push(&mut self, reg: Reg) -> u32 {
+        // The push is counted as pushing the 64-bit width in
+        // 64-bit architectures.
+        let size = 8u32;
+        self.reserve_stack(size);
+        let address = Address::from_shadow_sp(size as i64);
+        self.asm.str(reg, address, OperandSize::S64);
+
+        self.sp_offset
     }
+}
 
-    fn zero(&mut self, _reg: Reg) {
-        todo!()
+impl MacroAssembler {
+    fn increment_sp(&mut self, bytes: u32) {
+        self.sp_offset += bytes;
     }
 
-    fn push(&mut self, _reg: Reg) -> u32 {
-        todo!()
+    // Copies the value of the stack pointer to the shadow stack
+    // pointer: mov x28, sp
+
+    // This function is usually called whenever the real stack pointer
+    // changes, for example after allocating or deallocating stack
+    // space, or after performing a push or pop.
+    fn move_sp_to_shadow_sp(&mut self) {
```

Comment:
Could we add a block comment somewhere describing what the shadow stack is and how it works? (Or maybe I missed it earlier in which case let's have a reference to docs here!)

Reply:
I'd like to keep the documentation of the `shadow_sp` at `regs::shadow_sp()`, so I've added a reference here.